jsk_pcl_ros/src/kinfu_nodelet.cpp

```diff
@@ -33,236 +32,547 @@
+namespace pcl
+{
+  void
+  concatenateFields(PointCloud<PointXYZ>& cloud_xyz, PointCloud<RGB>& cloud_rgb, PointCloud<PointXYZRGB>& cloud)
+  {
+    if (cloud_xyz.points.size() != cloud_rgb.points.size())
+    {
+      std::cout << "Clouds being concatenated must have same size." << std::endl;
+      return;
+    }
+    cloud.points.resize(cloud_xyz.points.size());
+    for (size_t i = 0; i < cloud_xyz.points.size(); i++)
+    {
+      cloud.points[i].x = cloud_xyz.points[i].x;
+      cloud.points[i].y = cloud_xyz.points[i].y;
+      cloud.points[i].z = cloud_xyz.points[i].z;
+      cloud.points[i].r = cloud_rgb.points[i].r;
+      cloud.points[i].g = cloud_rgb.points[i].g;
+      cloud.points[i].b = cloud_rgb.points[i].b;
+    }
+    cloud.width = cloud_xyz.width;
+    cloud.height = cloud_xyz.height;
+  }
+}
 
 namespace jsk_pcl_ros
 {
-  void Kinfu::subscribe()
+  void
+  Kinfu::onInit()
   {
+    ConnectionBasedNodelet::onInit();
+    always_subscribe_ = true;  // for mapping
+
+    pnh_->param("device", device_, 0);
+    pnh_->param("auto_reset", auto_reset_, true);
+    pnh_->param("integrate_color", integrate_color_, false);
+    pnh_->param("slam", slam_, false);
+    pnh_->param<std::string>("fixed_frame_id", fixed_frame_id_, "odom_init");
+
+    pnh_->param<std::string>("save_dir", save_dir_, ".");
+    boost::filesystem::path save_dir(save_dir_);
+    if (boost::filesystem::create_directories(save_dir))
+    {
+      NODELET_INFO("Created save_dir: %s", save_dir_.c_str());
+    }
+
+    if (slam_)
+    {
+      tf_listener_.reset(new tf::TransformListener());
+    }
+
+    pub_camera_pose_ = advertise<geometry_msgs::PoseStamped>(*pnh_, "output", 1);
+    pub_cloud_ = advertise<sensor_msgs::PointCloud2>(*pnh_, "output/cloud", 1);
+    pub_generated_depth_ = advertise<sensor_msgs::Image>(*pnh_, "output/generated_depth", 1);
+    pub_rendered_image_ = advertise<sensor_msgs::Image>(*pnh_, "output/rendered_image", 1);
+
+    srv_reset_ = pnh_->advertiseService("reset", &Kinfu::resetCallback, this);
+    srv_save_mesh_ = pnh_->advertiseService("save_mesh", &Kinfu::saveMeshCallback, this);
 
+    onInitPostProcess();
   }
 
-  void Kinfu::unsubscribe()
+  void
+  Kinfu::initKinfu(const sensor_msgs::CameraInfo::ConstPtr& caminfo_msg)
   {
+    pcl::gpu::setDevice(device_);
+    pcl::gpu::printShortCudaDeviceInfo(device_);
+
+    /* below are copied from pcl/gpu/kinfu_large_scale/src/kinfuLS_app.cpp */
+
+    float vsz = pcl::device::kinfuLS::VOLUME_SIZE;
+    float shift_distance = pcl::device::kinfuLS::DISTANCE_THRESHOLD;
+    Eigen::Vector3f volume_size = Eigen::Vector3f::Constant(vsz/*meters*/);
+    if (shift_distance > 2.5 * vsz)
+    {
+      NODELET_WARN("WARNING Shifting distance (%.2f) is very large compared to the volume size (%.2f).",
+                   shift_distance, vsz);
+    }
+
+    kinfu_ = new pcl::gpu::kinfuLS::KinfuTracker(volume_size, shift_distance, caminfo_msg->height, caminfo_msg->width);
```

Comment:
I don't know C++ pointer well, but I think if kinfu is reset N times, new KinfuTracker object is generated N times without releasing memory. Is it my incorrect understanding?